py4cast/datasets/base.py

```diff
@@ -732,71 +733,71 @@ def cache_dir(self) -> Path:
-        )
         if self.settings.standardize:
             raise ValueError("Your dataset should not be standardized.")
-        # When computing stat may force every body to be input/ouput
-        for batch in tqdm(self.torch_dataloader(), desc="Computing stats"):
-
-            # Here we assume that data are in 2 or 3 D
-            inputs = batch.inputs.tensor
-            outputs = batch.outputs.tensor
-
-            # Check that no variable is a forcing variable
-            f_names = batch.forcing.feature_names
-            if len(f_names) > 0:
-                warnings.warn(
-                    f"Forcing variables {f_names} are present but no statistics will be computed."
-                )
 
-            in_out = torch.cat([inputs, outputs], dim=1)
-            in_out = in_out.flatten(
-                1, 3
-            )  # Flatten everybody to be (Batch, X, Features)
-
-            counter += in_out.shape[0]  # += batch size
+        for batch in tqdm(
+            self.torch_dataloader(), desc=f"Computing {type_tensor} stats"
+        ):
+            tensor = getattr(batch, type_tensor).tensor
+            tensor = tensor.flatten(1, 3)  # Flatten to be (Batch, X, Features)
+            counter += tensor.shape[0]  # += batch size
 
-            sum_means += torch.sum(in_out.mean(dim=1), dim=0)  # (d_features)
-            sum_squares += torch.sum((in_out**2).mean(dim=1), dim=0)  # (d_features)
+            sum_means += torch.sum(tensor.mean(dim=1), dim=0)  # (d_features)
+            sum_squares += torch.sum((tensor**2).mean(dim=1), dim=0)  # (d_features)
 
-            mini = torch.min(in_out, 1).values[0]
+            mini = torch.min(tensor, 1).values[0]
             stack_mini = torch.stack([best_min, mini], dim=0)
             best_min = torch.min(stack_mini, dim=0).values  # (d_features)
 
-            maxi = torch.max(in_out, 1).values[0]
+            maxi = torch.max(tensor, 1).values[0]
             stack_maxi = torch.stack([best_max, maxi], dim=0)
             best_max = torch.max(stack_maxi, dim=0).values  # (d_features)
 
         mean = sum_means / counter
         second_moment = sum_squares / counter
         std = torch.sqrt(second_moment - mean**2)  # (d_features)
 
-        # Storing variable statistics
-        store_d = {}
-        for i, name in enumerate(batch.inputs.feature_names):
-            store_d[name] = {
+        stats = {}
+        for i, name in enumerate(named_tensor.feature_names):
+            stats[name] = {
                 "mean": mean[i],
                 "std": std[i],
                 "min": best_min[i],
                 "max": best_max[i],
             }
+        return stats
+
+    def compute_parameters_stats(self):
+        """
+        Compute mean and standard deviation for this dataset.
+        """
+        all_stats = {}
+        for type_tensor in ["inputs", "outputs", "forcing"]:
+            stats_dict = self.compute_mean_std_min_max(type_tensor)
```

Comment:
The only problem I see here is that we loop 3 times over the whole dataset, once for inputs, once for outputs, once for forcing. 

Maybe the perfo overhead is ok, otherwise we might need to loop once and feep 3 stat aggregator objects in //

Reply:
I know it's not great, but it's something that we compute only once so...
And as I said above, when we refacto the dataset, we should put the stats computation inside the specific dataset class so it can be computed without using a dataloader and without making a difference between forcing, outputs, etc. The base class should not do this stuff.